frontend/pre_analysis.py

```diff
@@ -1,22 +1,196 @@
+        
+        Do the following:
+            - Propagate methods from bases to subclasses.
+            - Add __init__ function to classes if it doesn't exist.
+            - Return a mapping from class names to their attributes and methods.
+            - Return a mapping from class names to their base classes if they have any.
+            
+        """
+        # TODO propagate attributes to subclasses.
+        class_defs = [node for node in self.all_nodes if isinstance(node, ast.ClassDef)]
+
+        propagate_attributes_to_subclasses(class_defs)
+
+        class_to_attributes = OrderedDict()
+        class_to_base = OrderedDict()
+
+        for cls in class_defs:
+            if len(cls.bases) > 1:
+                raise NotImplementedError("Multiple inheritance is not supported yet.")
+            elif cls.bases:
+                class_to_base[cls.name] = cls.bases[0].id
+            else:
+                class_to_base[cls.name] = "object"
+
+            add_init_if_not_existing(cls)
+
+            attributes = set()
+            class_to_attributes[cls.name] = attributes
+
+            # Inspect all class-level statements
+            for cls_stmt in cls.body:
+                if isinstance(cls_stmt, ast.FunctionDef):
+                    # Add function to class attributes and get attributes defined by self.some_attribute = value
+                    attributes.add(cls_stmt.name)
+                    if not cls_stmt.args.args:
+                        continue
+                    first_arg = cls_stmt.args.args[0].arg  # In most cases it will be 'self'
+
+                    # Get attribute assignments where attribute value is the same as the first argument
+                    func_nodes = ast.walk(cls_stmt)
+                    assignments = [node for node in func_nodes if isinstance(node, ast.Assign)]
+
+                    for assignment in assignments:
+                        for target in assignment.targets:
+                            if (isinstance(target, ast.Attribute) and isinstance(target.value, ast.Name) and
+                                    target.value.id == first_arg):
+                                attributes.add(target.attr)
+                elif isinstance(cls_stmt, ast.Assign):
+                    # Get attributes defined as class-level assignment
+                    for target in cls_stmt.targets:
+                        if isinstance(target, ast.Name):
+                            attributes.add(target.id)
+
+        return class_to_attributes, class_to_base
+
+    def get_all_configurations(self):
+        config = Configuration()
+        config.max_tuple_length = self.maximum_tuple_length()
+        config.max_function_args = self.maximum_function_args()
+        config.classes_to_attrs, config.class_to_base = self.analyze_classes()
+        config.base_folder = self.base_folder
+
+        return config
+
+
+class Configuration:
+    """A class holding configurations given by the pre-analyzer"""
```

Comment:
For the entire program.

Because the constraints for the whole program (including the imported modules) are resolved at the end, so whole program analysis takes place first.